File: 07_linked_list/07_05_linked_list_cycle_II/linked_list_cycle_II.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def detect_cycle(head):
    slow = head
    fast = head
    while fast is not None and fast.next is not None:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            slow = head
            while slow != fast:
                slow = slow.next
                fast = fast.next
            return slow
    return None

# Test Case 1: Cycle present
head1 = ListNode(1)
node2 = ListNode(2)
node3 = ListNode(3)
node4 = ListNode(4)
node5 = ListNode(5)
head1.next = node2
node2.next = node3
node3.next = node4
node4.next = node5
node5.next = node2  # Creating a cycle
print("Test Case 1:", detect_cycle(head1).val)  # Output: 2

# Test Case 2: No cycle
head2 = ListNode(1, ListNode(2, ListNode(3)))
print("Test Case 2:", detect_cycle(head2))  # Output: None

# Test Case 3: Single node with a cycle
head3 = ListNode(1)
head3.next = head3  # Creating a cycle
print("Test Case 3:", detect_cycle(head3).val)  # Output: 1

# Test Case 4: Single node without a cycle
head4 = ListNode(1)
print("Test Case 4:", detect_cycle(head4))  # Output: None

# Test Case 5: Empty list
head5 = None
print("Test Case 5:", detect_cycle(head5))  # Output: None

File: 07_linked_list/07_05_linked_list_cycle_II/LinkedListCycleII.java
public class LinkedListCycleII {
    public static void main(String[] args) {
        // Test Case 1: Cycle present
        ListNode head1 = new ListNode(1);
        ListNode node2 = new ListNode(2);
        ListNode node3 = new ListNode(3);
        ListNode node4 = new ListNode(4);
        ListNode node5 = new ListNode(5);
        head1.next = node2;
        node2.next = node3;
        node3.next = node4;
        node4.next = node5;
        node5.next = node2; // Creating a cycle
        System.out.println("Test Case 1: " + detectCycle(head1).val); // Output: 2

        // Test Case 2: No cycle
        ListNode head2 = new ListNode(1, new ListNode(2, new ListNode(3)));
        System.out.println("Test Case 2: " + detectCycle(head2)); // Output: null

        // Test Case 3: Single node with a cycle
        ListNode head3 = new ListNode(1);
        head3.next = head3; // Creating a cycle
        System.out.println("Test Case 3: " + detectCycle(head3).val); // Output: 1

        // Test Case 4: Single node without a cycle
        ListNode head4 = new ListNode(1);
        System.out.println("Test Case 4: " + detectCycle(head4)); // Output: null

        // Test Case 5: Empty list
        ListNode head5 = null;
        System.out.println("Test Case 5: " + detectCycle(head5)); // Output: null
    }

    public static ListNode detectCycle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                slow = head;
                while (slow != fast) {
                    slow = slow.next;
                    fast = fast.next;
                }
                return slow;
            }
        }
        return null;
    }
}

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}